### PR TITLE
Fix [JENKINS-49206] by ensuring Timer threads get standard classloader

### PR DESCRIPTION
**Pull request by svanoort**
See [JENKINS-49206](https://issues.jenkins-ci.org/browse/JENKINS-49206).

### Proposed changelog entries

* Entry 1: JENKINS-49206, Ensure that Timer threads can't be created with a custom classloader - prevent possible Groovy memory leaks.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
### Desired reviewers

Test failure is unrelated -- @batmat might want to take a look at it, since it breaks for me without the patch.

Related tasks: 
- [x] When merged,  add a workflow-support note to switch to this the `Timeout` utility once that core is widely available.

@jglick

### CODE REVIEW

File: core/src/main/java/jenkins/util/Timer.java [hunk 1]
```diff
@@ -29,7 +29,19 @@
...
      * when running the test suite.
      */
-    private static ScheduledExecutorService executorService;
+    static ScheduledExecutorService executorService;
+
+
+    /** Avoids issues where the Timer threads are initialized in a context where they can receive a bogus contextClassLoader.
+     *  @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-49206">JENKINS-49206</>
+     */
+    static class ClassloaderSanityDaemonThreadFactory extends DaemonThreadFactory {
```
Comment (jglick) [inline core/src/main/java/jenkins/util/Timer.java]: Might want to make this an API in the same package as the others, using the delegate pattern like the others. Would be helpful e.g. for `workflow-support`’s `Timeout`.

Comment (jglick) [reply core/src/main/java/jenkins/util/Timer.java]: cf. https://github.com/jenkinsci/workflow-support-plugin/pull/53

Comment (svanoort) [reply core/src/main/java/jenkins/util/Timer.java]: That's a fair point

Comment (svanoort) [reply core/src/main/java/jenkins/util/Timer.java]: Done, @jglick

**Comment by batmat** [comment]
@svanoort care to clarify what you're referring to? Looking at CI I see a Javadoc failure and no test failure? (Though I'm otp so possibly mislooking).

**Comment by svanoort** [comment]
Weird.  Okay I was seeing a permissions failure locally with the AtomicFileWriterTest @batmat  -- but it may be masked by whatever is hosed with the core Javadocs.

**Comment by svanoort** [comment]
@batmat  This is the one:

> ERROR] checkPermissions(hudson.util.AtomicFileWriterTest)  Time elapsed: 0.043 s  <<< FAILURE!
java.lang.AssertionError: missing: GROUP_WRITE
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at hudson.util.AtomicFileWriterTest.checkPermissions(AtomicFileWriterTest.java:187)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)

Failing locally on Mac -- trying to debug it, I discovered that it is ONLY GROUP_WRITE missing for me. Maybe a Mac-specific impl quirk?

**Review by oleg-nenashev — APPROVED** [review]
🐝 AFAICT. Same probably needs to be applied to other executor services

**Comment by batmat** [comment]
@svanoort your issue was real fwiw, fixed in https://github.com/jenkinsci/jenkins/pull/3275

**Review by KostyaSha — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@reviewbybees done, I'd guess

File: core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package hudson.util;
```
Comment (jglick) [inline core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java]: license header

File: core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.concurrent.TimeUnit;
+
+/**
+ *  Explicitly sets the {@link Thread#contextClassLoader} for threads it creates to its own classloader.
+ *  This avoids issues where threads are lazily created (ex by invoking {@link java.util.concurrent.ScheduledExecutorService#schedule(Runnable, long, TimeUnit)})
+ *   in a context where they would receive a customized {@link Thread#contextClassLoader} that was never meant to be used.
+ *
+ *  Commonly this is a problem for Groovy use, where this may result in memory leaks.
+ *  @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-49206">JENKINS-49206</a>
+ */
```
Comment (jglick) [inline core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java]: `@since FIXME`

File: core/src/main/java/jenkins/util/Timer.java [hunk 4]
```diff
@@ -29,7 +30,8 @@
      * The scheduled executor thread pool. This is initialized lazily since it may be created/shutdown many times
      * when running the test suite.
      */
-    private static ScheduledExecutorService executorService;
+    static ScheduledExecutorService executorService;
```
Comment (jglick) [inline core/src/main/java/jenkins/util/Timer.java]: Revert, apparently unused in test.

File: core/src/main/java/jenkins/util/Timer.java [hunk 5]
```diff
@@ -42,7 +44,7 @@ public static synchronized ScheduledExecutorService get() {
             // corePoolSize is set to 10, but will only be created if needed.
             // ScheduledThreadPoolExecutor "acts as a fixed-sized pool using corePoolSize threads"
             // TODO consider also wrapping in ContextResettingExecutorService
-            executorService = new ImpersonatingScheduledExecutorService(new ErrorLoggingScheduledThreadPoolExecutor(10, new NamingThreadFactory(new DaemonThreadFactory(), "jenkins.util.Timer")), ACL.SYSTEM);
+             executorService = new ImpersonatingScheduledExecutorService(new ErrorLoggingScheduledThreadPoolExecutor(10, new NamingThreadFactory(new ClassLoaderSanityThreadFactory(new DaemonThreadFactory()), "jenkins.util.Timer")), ACL.SYSTEM);
```
Comment (jglick) [inline core/src/main/java/jenkins/util/Timer.java]: extra indentation

File: core/src/test/java/jenkins/util/TimerTest.java [hunk 6]
```diff
@@ -44,4 +50,52 @@ protected void doRun() throws Exception {
...
+    @Issue("JENKINS-49206")
+    public void timerBogusClassloader() throws Exception {
+        final int threadCount = 10;  // Twice Timer pool size to ensure we end up creating at least one new thread
+        final CountDownLatch startLatch = new CountDownLatch(threadCount);
+
+        final ClassLoader[] contextClassloaders = new ClassLoader[threadCount];
+        ScheduledFuture[] futures = new ScheduledFuture[threadCount];
+        final ClassLoader bogusClassloader = new GroovyClassLoader();
+
+        Runnable timerTest = new Runnable() {
```
Comment (jglick) [inline core/src/test/java/jenkins/util/TimerTest.java]: could use a lambda BTW

**Review by jglick — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
As discusses with @svanoort, merging as is to get it in the weekly. Code formatting can be hanled as a follow-up

File: core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java (hunk 3 above)
Comment (basil) [reply core/src/main/java/hudson/util/ClassLoaderSanityThreadFactory.java:13]: Javadoc links were invalid references: https://github.com/jenkinsci/jenkins/pull/6710#pullrequestreview-1021860493

**Pull request merged**